### ISSUE TITLE
Adaptive choice of compaction algorithms

### ISSUE THREAD

**Issue by kjosib** [kjosib/booze-tools#11]
The GOTO compaction algorithm seems to be about optimal over a wide range of grammars, but the DFA and ACTION tables seem to display different characteristics depending on the size and complexity of the input grammar and scanner definitions. It may be worthwhile for the compiler to try a couple different approaches and see which method yields the smallest tables -- or perhaps let the user select a choice of table storage method at compile-time. The downside is that a complete port of the run-time would have to account for the variation inherent in the system. On the other hand, choice of compaction methods can also influence the size/speed trade-off, or provide an anchor-point for organizing a survey of different methods that have been used in past decades.

**Comment by kjosib** [comment]
Reference:
https://www.researchgate.net/publication/220404687_Optimization_of_Parser_Tables_for_Portable_Compilers

TLDR: They tested various table compaction methods, alone and in combination, for 11 or so grammars looking at the time/space trade-off. They mention a source for published grammars, but it seems to be self-referential back to the same paper.
